Better trigger jump fix

jason-e opened the pull request:
I tested `sv_ground_trigger_fix` and noticed it wasn't catching all cases of the bug so I made some changes to make it more accurate.

The "jumpbug" fix wasn't always correctly determining if the player would unduck, and I figured out that splitting the trace it was doing into two (one for unducking, one for landing) fixed this, though I'm not entirely sure why (trace shenanigans?). It also uses the exact same logic as `CanUnduck()` now, which should make it even more accurate.

The other part of the fix is the trigger jumping bug where you can land at the end of one tick and jump the next without activating triggers due to `sv_considered_on_ground` keeping you above them. This fix needs to know where the ground actually is below you, which depends on the logic in `CategorizePosition()`, some of which was not already accounted for in Momentum's fix. Rather than re-implement more of that function, I just saved off the known ground position when `SetGroundEntity()` is called and used that value to do the fix, which means that fix should have less (if any) false negatives now while also being cleaner.

I also noticed that the jumpbug fix calls `ProcessImpacts()` directly, which means it can be called twice in one tick. Unfortunately this means that some trigger `Touch()` functions may be called twice in a tick, and in the case of `trigger_push` entities, it can result in double boosts as they are designed to only be called once per tick. I considered making touch system changes to keep `Touch()` functions from being called twice in a tick, but that would be somewhat tricky and I am not sure if that would have side effects. `trigger_push` entities are the only thing I personally know of that depends on one `Touch()` call per tick, so instead I added a very simple fix specifically targeting them.


### Checklist
- [x] **I have thoroughly tested all of the code I have modified/added/removed to ensure something else did not break**
- [x] If there is a localization token change, I have updated the `momentum_english_ref.res` file with the changes, ran `tokenizer.py` to generate an up-to-date localization file, and have committed both the `.res` file changes and the new localization `.txt` file
- [x] If I introduced new h/cpp files, I have added them to the appropriate project's VPC file (`server_momentum.vpc` / `client_momentum.vpc` / etc)
- [x] If I have added or modified any visual assets (models, materials, panels, effects, etc), I have taken screenshots / videos of them and attached them to this PR directly (screenshots uploaded through github, videos uploaded to youtube and linked)
- [x] If I have modified any console command, console variable, or momentum entity, I have opened an issue (or a PR) for it in the [Momentum Mod documentation repository](https://github.com/momentum-mod/docs)
- [x] My commits are relatively small and scoped to the best of my ability
- [x] My branch has a clear history of changes that can be easy to follow when being reviewed commit-by-commit
- [x] My branch is functionally complete; the only changes to be done will be those potentially requested in code review

<!-- If any of these items are giving you doubts, please ask about it in the Discord! -->

jason-e committed:
Improve ground trigger fixes accuracy

- Split unduck and land checks into separate traces to fix false-negative unduck checks
- Use complete CanUnduck logic to make sure the player actually will unduck
- For the end of tick fix, remember the actual ground position from ProcessMovement instead of finding the ground again with approximated logic

jason-e committed:
Prevent jumpbug fix from causing double boosts

chrb22 commented:
So from what I understand, the original implementation for jumpbugs thought you couldn't unduck while you actually could? I also noticed the double triggering bug and I have been able to make a hacky fix for it, but I haven't pushed it as I kind of wanted to revamp the bounce prevention at the same time for whenever I make the bounce trigger, possibly together with a version of `sv_edge_fix` that accounts for triggers.

jason-e commented:
https://user-images.githubusercontent.com/2745352/121823804-1cb41e80-cc6d-11eb-8893-32bf1844dd42.mp4

For some reason it was just barely thinking I wouldn't be able to unduck when I actually could. It would say the ground was something like 8.495 units away when it was actually exactly 8.5 away, which is a scenario that happens in Momentum simply by jumping on flat ground (100 tick). Rather than do the unduck distance check and the landing height check with one trace, I split it into two and it fixed it (and I also changed the unduck logic to match exactly how CanUnduck works).

What did you have in mind for fixing the double trigger issue? Did you notice it causing problems with just `trigger_push` or is there another entity that needs to be accounted for?

chrb22 commented:
Okay, then I know what the problem you found is. Swept traces return a position that is generally 0.03125 units (`DIST_EPSILON`) away from a surface. Can't remember if I didn't account for it because I didn't know about it yet when I wrote it or it simply didn't come to mind.

I didn't know of any triggers myself that could be affected by the double triggering, but I just kind of assumed there would be. My solution is to check if the trigger `touchStamp` was updated when `ProcessImpacts` was called last tick and use that to filter triggers that should not be activated again in the new tick. As the jumpbug check can be thought of as a continuation of the previous tick, I simply don't update the player `touchStamp` in `PhysicsTouchTriggers -> PhysicsTouchTriggers -> SetCheckUntouch` in that case, such that any trigger with an updated `touchStamp` from the previous tick won't get activated. `CCollisionProperty::CheckForUntouch` will also update the player `touchStamp` when they are teleported, but that only happens if they are hugging are wall or similar, so I ended up ignoring players in that function as it would affect the implementation.

jason-e commented:
Here's the exact output of an example where `PreventBounce` was failing:
```
jumpbug PreventBounce trace hit -> start z: 8.526255, end z: 0.031250, dist: 8.495005, nrm: 1.000000, can unduck: false
CanUnduck trace -> fraction: 1.000000, start z: 8.526253, end z: 0.026253, dist: 8.500000, can unduck: true
```
It seems perhaps that if a swept trace can reach `fraction == 1.0` without hitting something, the exact end point is returned, but if it does hit something then its end position is rounded to `FLT_EPSILON`? If that is the case then having this fix use two separate traces is definitely the way to go, especially since it keeps things 1:1 with how `ProcessMovement()` works so there's no doubt about its accuracy. Ideally the fix would actually call an abstracted version of `CanUnduck()` directly, but I stopped short of setting that up myself.

For the double triggering, are you saying you would change the extra `ProcessImpacts()` call being made at the beginning of the tick to not increment `touchStamp`? I think that sounds like a decent general solution. One tricky thing is the fact that the tick lifecycle for `basevelocity` starts/ends with `CheckMovingGround()` rather than `ProcessMovement()`. If we did make that change to the behavior of `touchStamp` for our extra `ProcessImpacts()` call, we would also have to move that call to before `CheckMovingGround()`, which would also put it before `PreThink()` and `Think()`. I'm not super familiar with the importance of those calls for players, so I would be wary of creating different side effects by moving things, though it could be fine. I guess it comes down to whether or not we think we need a more-involved generalized fix for double-touching, or if this pull request's simple (if hacky) fix for just `trigger_push` entities is all that is needed.

chrb22 commented:
That's exactly how swept traces work, so your edit is definitely necessary.

I am not exactly sure myself how base velocity is used and handled in the code, so I can't really comment on that. When I thought of double triggering, it was mostly for maps that might use more complicated systems. For example Gorge004 makes a lot of [crazy gimmicks](https://www.youtube.com/watch?v=jKmGKsQiWiE) and I assume some of them rely on a steady 1 activation per tick.

jason-e commented:
I took a quick look at all of the trigger classes and their Touch functions and it looks like there's at least a couple where extra Touch calls could be problematic (`trigger_catapault` and `trigger_momentum_userinput`). Both of those have entity IO events that fire when the Touch function does it's thing. Also, with `trigger_catapault` it might be possible to get more of a boost by abusing the double Touch but I'm not sure as I haven't played games with that entity so I don't fully understand how it works. 

That being said, with those entities (and honestly even `trigger_push`) I think the odds of everything aligning in a way that leads to double Touch abuse even being possible is extremely unlikely. Personally, I'm not going to worry about messing with overall Touch system stuff just yet. It's a wide reaching system and I don't want to mess something up to fix a bug that might not matter... unless perhaps I'm underestimating `trigger_catapault` entities since I havent played jump maps.

hexaflexahexagon commented:
The way trigger_catapult works is it sets your velocity to X, so doubleboosting is irrelevant. Contrasting with trigger_push which just adds X to your velocity.

e: and here is userinput. https://docs.momentum-mod.org/entity/trigger_momentum_userinput/
haven't use this one myself but it also doesn't seem like an issue

jason-e commented:
I noticed that trigger_catapault's "useExactVelocity" setting involves checking the player's current velocity at some point, though without sitting down and working through all the math there I don't know what that is actually doing.

The userinput trigger fires OnKeyPressed / OnKeyHeld / OnKeyReleased each time Touch() is called and those events' criteria are met. Those outputs could be linked to literally anything by a mapper and those things might depend on one call per tick. Catapaults also fire similar outputs when activated (even if they just set your velocity to what it already was).

Though again, the odds of a trigger being positioned and shaped in a way that it is double-touchable, AND those outputs are linked to something that depends on one call per tick are probably unlikely.

chrb22 commented:
I think this is fine for now. Using touchstamps would be the best solution, but I need to dig deeper into it to ensure it doesn't affect anything else.

lvaness commented:
With regards to the catapult, keep in mind we now use Portal 2's trigger_catapult (used in "Aerial Faith Plates" there)
